tests/test_mark_tokens.py

```diff
@@ -338,6 +339,39 @@ def test_adjacent_joined_strings(self):
             "JoinedStr:'x y z'   # comment2\n       'a b c'   # comment3\n       f'u v w'"
         })
 
+  if sys.version_info >= (3, 12):
+    def test_fstrings_3_12_plus(self):
+        m = self.create_mark_checker(
+          'x = (f"Wobble {f"{func(kwarg=f"{boo!r}")}"!r}.",)',
+        )
+
+        self.assertEqual(m.view_nodes_at(1, 6), {
```

Comment:
Can you also check one or both of the inner f-strings?

Reply:
https://github.com/gristlabs/asttokens/pull/134/commits/1c42c835b0bb525afd2832478c4bf41640027b87 does this to an extent. We don't seem to get the full details of the inner nodes though. I'm not sure if that's expected.

Reply:
Ah, that's sort of expected pre-3.12 but it's also sort of a bug in 3.12 where getting the tokens becomes possible. ASTText can deal with f-strings, but the package is called asttokens...so if you're willing to look into this, search for `include_joined_str`.

Reply:
Yeah, I'd seen that argument to the `iter_children_*` functions. That's the reason for using `node.values` over the `view_nodes_at` approach in the below assertion. Perhaps I'm misunderstanding, but I think the lack of text is coming from the calls to `ASTTokens.get_text_positions` where the implementations don't return useful values for f-strings (due to not having a `first_token`). Checking `ASTText.get_text_positions`'s behaviour might be useful, though I'm assuming that that's covered elsewhere.

I'm not sure what you're suggesting be investigated/potentially changed here?

Reply:
I'm suggesting that in 3.12+, `MarkTokens` should start attaching `first_token` to inner f-string nodes.

Reply:
Ah, I see. That feels worthwhile, but perhaps a bit out of the scope of this PR? My main goal here was to add some lightweight coverage for things that #132 seemed to show weren't covered by unit tests.

Having a very quick look, just adding `include_joined_str=True` in a few places doesn't seem to pass the tests, so this would probably need a deeper dive really.

Reply:
Right, please just add a comment explaining the situation with a TODO, because the assertion is indeed hard to understand.

Reply:
Is 736d4e3 sufficiently clear?